### PR TITLE
[Forge] Add Full node TPS test

### PR DESCRIPTION
**Pull request by sitalkedia**
### Description

Add a performance test that spawns 5 FNs and 30 validators and sends transactions only to the FNs. This will give us idea around mempool performance. 

### Test Plan

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2553)

### CODE REVIEW

File: testsuite/run_forge.sh [hunk 1]
```diff
@@ -42,7 +42,7 @@ FORGE_RUNNER_MODE=${FORGE_RUNNER_MODE:-k8s}
 FORGE_NAMESPACE_KEEP=${FORGE_NAMESPACE_KEEP:-false}
 FORGE_NAMESPACE_REUSE=${FORGE_NAMESPACE_REUSE:-false}
 FORGE_ENABLE_HAPROXY=${FORGE_ENABLE_HAPROXY:-false}
-FORGE_TEST_SUITE=${FORGE_TEST_SUITE:-land_blocking}
+FORGE_TEST_SUITE=${FORGE_TEST_SUITE:-bench_with_fullnode}
```
Comment (sitalkedia) [inline testsuite/run_forge.sh:45]: This is temp change - will revert once we run forge on this PR.

**Comment by bchocho** [comment]
Comparing this run with a recent regular run:
* [this run, only looking at FNs](https://o11y.aptosdev.com/grafana/d/mempool-brian-updated/mempool?from=1659667927000&to=1659669136000&var-Datasource=Remote%20Prometheus%20Devinfra&var-namespace=forge-pr-2553&var-chain_name=forge-0&var-kubernetes_pod_name=All&orgId=1&var-role=fullnode&var-interval=$__auto_interval_interval)
* [recent run, validators](https://o11y.aptosdev.com/grafana/d/mempool-brian-updated/mempool?from=1659677617000&to=1659678410000&var-Datasource=Remote%20Prometheus%20Devinfra&var-namespace=forge-pr-2537&var-chain_name=forge-0&var-kubernetes_pod_name=All&orgId=1)

One thing that stands out immediately is mempool size is low for some nodes, and new transactions received rate is too low to get anything near 7K TPS. If we tuned txn emitter could we get higher numbers?

**Comment by sitalkedia** [comment]
> Comparing this run with a recent regular run:
> 
> * [this run, only looking at FNs](https://o11y.aptosdev.com/grafana/d/mempool-brian-updated/mempool?from=1659667927000&to=1659669136000&var-Datasource=Remote%20Prometheus%20Devinfra&var-namespace=forge-pr-2553&var-chain_name=forge-0&var-kubernetes_pod_name=All&orgId=1&var-role=fullnode&var-interval=$__auto_interval_interval)
> * [recent run, validators](https://o11y.aptosdev.com/grafana/d/mempool-brian-updated/mempool?from=1659677617000&to=1659678410000&var-Datasource=Remote%20Prometheus%20Devinfra&var-namespace=forge-pr-2537&var-chain_name=forge-0&var-kubernetes_pod_name=All&orgId=1)
> 
> One thing that stands out immediately is mempool size is low for some nodes, and new transactions received rate is too low to get anything near 7K TPS. If we tuned txn emitter could we get higher numbers?

@bchocho - Yes the TPS is lower not because the emitter is not tuned well enough but because the transaction submission from the REST API becomes the bottleneck. In this case, we have just 6 REST end point vs 30. tinyurl.com/2cl42ube shows the REST API latency for validator test  and tinyurl.com/29mtcatf is the REST end point latency for FN test. You will see the REST latency is ~50 ms in case of validator test and is upto 1s in this case. Let me try increasing the FNs to 30 to see how that behaves.

**Comment by bchocho** [comment]
Sorry, I should have probably expanded more on `new transactions received rate` a bit. Late night...
* Run w/o FNs: I see each mempool is consistently getting 200-250 txn/s on their REST endpoint ('client network') in the chart. Because there are 30 endpoints, 233 txn/s * 30 = 7K txn/s.
* Run w/ 6 FNs: Each FN mempool does get > 250 txn/s but to achieve the same TPS they would have to get ~1.2K txn/s.

I was suspecting that we just weren't sending enough txns at the txn emitter, but seems from your observations on latency, it looks like we are hitting a limit on REST request processing which we'll have to look into. Hopefully something simple like "add an extra thread" :P

BTW, in forge are we using FNs that are as powerful as the validator nodes?

**Comment by sitalkedia** [comment]
> BTW, in forge are we using FNs that are as powerful as the validator nodes?

Yes, that's the case AFAIK. @rustielin can confirm.

**Comment by rustielin** [comment]
> BTW, in forge are we using FNs that are as powerful as the validator nodes?

Oh nice catch, no actually. https://github.com/aptos-labs/internal-ops/pull/368 this would fix that.

**Comment by sitalkedia** [comment]
@bchocho, @rustielin - Can I get a stamp on this? Let's land this change and we can iterate on it to experiments with various configurations for FNs to achieve higher TPS.

File: testsuite/run_forge.sh (hunk 1 above)
Comment (bchocho) [reply testsuite/run_forge.sh:45]: Assume this will be reverted before land?

**Review by bchocho — APPROVED** [review]

**Pull request merged**